Fix Notepad commandline style parsing

CookiePLMonster opened the pull request:
This PR refines `-notepadStyleCmdline` parsing to e.g. properly account for filenames which have multiple spaces in the file name.

To verify, consider this case:
1. Have Notepad++ replace Notepad, as described here: https://notepad-plus-plus.org/features/replace-notepad.html
2. Create a file called `file name.txt` (with one space).
3. Create a file called `file  name.txt` (with two spaces).
4. In the directory containing those files, run `notepad file  name` (with one space).

### Observed result

`file name.txt` (with one space) is opened in Notepad++.

### Expected result

`file  name.txt` (with two spaces) is opened in Notepad++.

Fixes #5861

madivad commented:
I'm slightly confused why you would expect the filename with two spaces to be opened when you'e specifically asked for one. 

If you had of designed the test for failure in the opposite direction (asking for the filename with two spaces and only getting the one with ONE space, THEN I'd understand). Can you elaborate?

CookiePLMonster commented:
> I'm slightly confused why you would expect the filename with two spaces to be opened when you'e specifically asked for one.

GitHub folds those multiple spaces into one, hence me commenting those with (one space) and (two spaces).


> perusing this code, why is the command line tokenised to begin with? Why not accept the filenames as presented by the operating system?

It's tokenized as if it was passed as `argc` and `argv`. Notepad++ has numerous commandline arguments, so this is pretty much a necessity.

And no, this very pull request addresses issues occuring when a filename is passed **not** escaped. You may consider this a bogus use case, but Windows Notepad handles it correctly (as per my notes), and the entire purpose of `-notespadStyleCmdline` switch is to support Notepad commandline arguments perfectly, no matter how weird they may be.

xylographe commented:
> GitHub folds those multiple spaces into one

Not if you use U+00A0 (NO-BREAK SPACE),
 `file  name.txt` or
 `file name.txt`. :)

AngryGamer commented on PowerEditor/src/winmain.cpp:
```diff
@@ -101,6 +105,21 @@ void stripIgnoredParams(ParamVector & params)
+					// Advance to the first non-whitespace and not quotation mark character
+					while ( iswspace( *ignoredString ) || *ignoredString == L'"' )
+					{
+						if ( *ignoredString == '\0' )
```
Does this do anything? I don't see either while condition letting a NULL character through.

CookiePLMonster replied:
Good point - according to https://en.cppreference.com/w/cpp/string/wide/iswspace null terminator is not considered whitespace, and thus it should be out of this loop by then. Will correct tomorrow.

AngryGamer commented on PowerEditor/src/winmain.cpp:
```diff
@@ -110,41 +129,37 @@ void stripIgnoredParams(ParamVector & params)
-		ssHasContents = true;
-		ss << *it;
-		if ( std::next(it) != params.end() ) ss << TEXT(" ");
+		if ( *pCmdLine == '\0' )
```
ditto

donho commented on PowerEditor/src/winmain.cpp:
```diff
@@ -313,10 +328,10 @@ void doException(Notepad_plus_Window & notepad_plus_plus)
 
 
-int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE, LPSTR, int)
+int WINAPI wWinMain(HINSTANCE hInstance, HINSTANCE, PWSTR pCmdLine, int)
```
Why do we need `wWinMain` ?

CookiePLMonster replied:
I need access to a wide char command line parameter. `GetCommandLineW` does fit my use case, because it also includes the program name. See:

```
lpCmdLine

Type: LPSTR

The command line for the application, excluding the program name. To retrieve the entire command line, use the GetCommandLine function.
```

CookiePLMonster commented:
Oh my, oh my! I approached this code intending to resolve concerns raised in a review and I was in shock. Have I even tested this!?

First of all, this commit appears to be bigger than it actually is, because I had to move `stripIgnoredParams` elsewhere. Here's what I fixed, though:
1. `-notepadStyleCmdline` was not stripped from a joint filename path, only `-z` was That means `notepad++.exe -notepadStyleCmdline file name.txt` attempted to open `-notepadStyleCmdline file name.txt`. Now fixed.
2. `.txt` was appended even if there was no filename. That means `notepad++.exe -notepadStyleCmdline` attempted to open `.txt`. Now fixed.
3. Skipping over commandline parameters was not checking if entire words were matched. This means `notepad++.exe filenamewith-z-inparam -z paramtoskip` would be totally broken - when attempting to skip `-z`, it matched the one in `filenamewith-z-inparam`. Now fixed.

This is pretty embarrassing. I hope it's all fine now.

donho commented:
@CookiePLMonster  Indeed, I was thinking "This function is quite complicated for the divers features, let's uni-test it. to avoid the regression". I installed gTest via Visual Studio and begin to do some tests, then I realized the test part should be a lib or a dll, so I quit. If there's the unit-test framework which doesn't need to modify the structure of Notepad++ part. I'm all for unit-test but I hate to modify the main project's structure for adapting the tests.

donho commented:
Tried also catch2 without success. I quit for unit-test for the moment.
